hail/src/main/scala/is/hail/expr/ir/agg/CallStatsAggregator.scala

```diff
@@ -0,0 +1,215 @@
+import is.hail.io.{CodecSpec, InputBuffer, OutputBuffer}
+import is.hail.stats.CallStats
+import is.hail.utils._
+
+
+object CallStatsState {
+  val callStatsInternalArrayType = PArray(PInt32Required, required = true)
+  val stateType: PTuple = PTuple(true, callStatsInternalArrayType, callStatsInternalArrayType)
+
+  val resultType = PStruct(
+    "AC" -> PArray(PInt32()),
+    "AF" -> PArray(PFloat64()),
+    "AN" -> PInt32(),
+    "homozygote_count" -> PArray(PInt32()))
+
+  assert(resultType.virtualType == CallStats.schema)
+}
+
+case class CallStatsState(mb: EmitMethodBuilder) extends PointerBasedRVAState {
+  val alleleCountsOffset: Code[Long] = CallStatsState.stateType.fieldOffset(off, 0)
+  val homCountsOffset: Code[Long] = CallStatsState.stateType.fieldOffset(off, 1)
+  val alleleCounts: Code[Long] = CallStatsState.stateType.loadField(off, 0)
+  val homCounts: Code[Long] = CallStatsState.stateType.loadField(off, 1)
+  val nAlleles: Code[Int] = CallStatsState.callStatsInternalArrayType.loadLength(alleleCounts)
+  private val addr = mb.newField[Long]
+
+  // unused but extremely useful for debugging if something goes wrong
+  def dump(tag: String): Code[Unit] = {
+    val i = mb.newField[Int]
+    Code(
+      Code._println(s"at tag $tag"),
+      i := 0,
+      Code.whileLoop(i < nAlleles,
+        Code._println(const("at i=").concat(i.toS).concat(", AC=").concat(alleleCountAtIndex(i, nAlleles).toS).concat(", HOM=").concat(homCountAtIndex(i, nAlleles).toS)),
+        i := i + 1)
+    )
+  }
+
+  def alleleCountAtIndex(idx: Code[Int], length: Code[Int]): Code[Int] =
+    Region.loadInt(CallStatsState.callStatsInternalArrayType.loadElement(region, alleleCounts, length, idx))
+
+  def updateAlleleCountAtIndex(idx: Code[Int], length: Code[Int], updater: Code[Int] => Code[Int]): Code[Unit] = Code(
+    addr := CallStatsState.callStatsInternalArrayType.loadElement(region, alleleCounts, length, idx),
+    Region.storeInt(addr, updater(Region.loadInt(addr)))
+  )
+
+  def homCountAtIndex(idx: Code[Int], length: Code[Int]): Code[Int] =
+    Region.loadInt(CallStatsState.callStatsInternalArrayType.loadElement(region, homCounts, length, idx))
+
+
+  def updateHomCountAtIndex(idx: Code[Int], length: Code[Int], updater: Code[Int] => Code[Int]): Code[Unit] = Code(
+    addr := CallStatsState.callStatsInternalArrayType.loadElement(region, homCounts, length, idx),
+    Region.storeInt(addr, updater(Region.loadInt(addr)))
+  )
+
+  def serialize(codec: CodecSpec): Code[OutputBuffer] => Code[Unit] = {
+    codec.buildEmitEncoderF[Long](CallStatsState.stateType, CallStatsState.stateType, mb.fb)(region, off, _)
+  }
+
+  def deserialize(codec: CodecSpec): Code[InputBuffer] => Code[Unit] = {
+    off := codec.buildEmitDecoderF(CallStatsState.stateType, CallStatsState.stateType, mb.fb)(region, _)
+  }
+
+  def copyFromAddress(src: Code[Long]): Code[Unit] = {
+    StagedRegionValueBuilder.deepCopy(er, CallStatsState.stateType, region.loadAddress(src), off)
```

Comment:
this doesn't need the `region.loadAddress` around `src`---that's what the `copyFrom` wrapper will do

Reply:
ahh, got it